contrib/sarplus/python/pysarplus/SARPlus.py

```diff
@@ -488,3 +524,72 @@ def recommend_k_items(
             )
         else:
             raise ValueError("No cache_path specified")
+
+    def get_topk_most_similar_users(self, test, user, top_k=10):
+        """Based on user affinity towards items, calculate the most similar
+            users to the given user.
+
+        Args:
+            test (pyspark.sql.DataFrame): test Spark dataframe.
+            user (int): user to retrieve most similar users for.
+            top_k (int): number of top items to recommend.
+
+        Returns:
+            pyspark.sql.DataFrame: Spark dataframe with top k most similar users
+            from test and their similarity scores in descending order.
+        """
+
+        if len(test.filter(test["user_id"].contains(user)).collect()) == 0:
+            raise ValueError("Target user must exist in the input dataframe")
+
+        test_affinity = self.get_user_affinity(test).alias("matrix")
+        num_test_users = test_affinity.select("user_id").distinct().count() - 1
+
+        if num_test_users < top_k:
+            log.warning(
+                "Number of users is less than top_k, limiting top_k to number of users"
+            )
+        k = min(top_k, num_test_users)
+
+        user_affinity = test_affinity.where(F.col("user_id") == user).alias("user")
+
+        df_similar_users = (
+            test_affinity.join(
+                user_affinity,
+                test_affinity["item_id"] == user_affinity["item_id"],
+                "outer",
+            )
+            .withColumn(
+                "prod",
+                F.when(F.col("matrix.user_id") == user, -float("inf"))
+                .when(
+                    F.col("user.rating").isNotNull(),
+                    F.col("matrix.rating") * F.col("user.rating"),
+                )
+                .otherwise(0.0),
+            )
+            .groupBy("matrix.user_id")
+            .agg(F.sum("prod").alias("similarity"))
+            .orderBy("similarity", ascending=False)
+            .limit(k)
+        )
+
+        return df_similar_users
+
+    def get_popularity_based_topk(self, top_k=10, items=True):
+        """Get top K most frequently occurring items across all users.
+
+        Args:
+            top_k (int): number of top items to recommend.
+            items (bool): if false, return most frequent users instead.
+
+        Returns:
+            pyspark.sql.DataFrame: Spark dataframe with top k most popular items
+            and their frequencies in descending order.
+        """
+
+        # TODO: get most frequent users
+        if not items:
+            raise ValueError("Not implemented")
+
+        return self.item_frequencies.orderBy("frequency", ascending=False).limit(top_k)
```

Comment:
I noticed that the sarplus tests are triggered, but there is a badge in the main page that is failing: https://github.com/microsoft/recommenders/tree/staging/contrib/sarplus @simonzhaoms do you know what is the problem?